File: SlooshIOS/neomovies-android-main/app/src/main/java/com/neo/neomovies/data/alloha/AllohaRepository.kt
package com.neo.neomovies.data.alloha

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.net.URLEncoder
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

private const val TAG = "AllohaRepository"

data class AllohaTranslation(
    val id: String,
    val name: String,
    val iframeUrl: String,
)

data class AllohaEpisode(
    val season: Int,
    val episode: Int,
    val translations: List<AllohaTranslation>,
)

data class AllohaSeason(
    val season: Int,
    val episodes: List<AllohaEpisode>,
)

data class AllohaMovie(
    val title: String,
    val iframeUrl: String,
    val translations: List<AllohaTranslation>,
)

data class AllohaApiResult(
    val title: String,
    val isSerial: Boolean,
    val movie: AllohaMovie?,
    val seasons: List<AllohaSeason>,
)

class AllohaRepository(
    private val okHttpClient: OkHttpClient,
    private val token: String = "ffbd312217e27c4245f2678afe1881",
) {

    private fun buildTrustingClient(): OkHttpClient {
        val trustAllCerts = arrayOf<TrustManager>(object : X509TrustManager {
            override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
            override fun checkClientTrusted(certs: Array<X509Certificate>?, authType: String?) {}
            override fun checkServerTrusted(certs: Array<X509Certificate>?, authType: String?) {}
        })
        val sc = SSLContext.getInstance("TLS")
        sc.init(null, trustAllCerts, SecureRandom())
        return okHttpClient.newBuilder()
            .sslSocketFactory(sc.socketFactory, trustAllCerts[0] as X509TrustManager)
            .hostnameVerifier { _, _ -> true }
            .build()
    }

    suspend fun fetchByKpId(kpId: Int): AllohaApiResult = withContext(Dispatchers.IO) {
        val encodedToken = URLEncoder.encode(token, "UTF-8")
        val encodedKp = URLEncoder.encode(kpId.toString(), "UTF-8")
        val apiUrl = "https://api.alloha.tv/?token=$encodedToken&kp=$encodedKp"

        Log.d(TAG, "fetchByKpId: kpId=$kpId url=$apiUrl")

        val client = buildTrustingClient()
        val request = Request.Builder().url(apiUrl).build()
        val jsonStr = retry(times = 3, delayMs = 1000L) {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw RuntimeException("Alloha API returned ${response.code}")
                }
                response.body?.string() ?: throw RuntimeException("Empty response body")
            }
        }

        val dataObj = JSONObject(jsonStr).getJSONObject("data")
        val title = dataObj.optString("name", "Unknown")
        val seasonsObj = dataObj.optJSONObject("seasons")

        if (seasonsObj != null) {
            val parsedSeasons = mutableListOf<AllohaSeason>()
            seasonsObj.keys().forEach { sKey ->
                val sObj = seasonsObj.getJSONObject(sKey)
                val episodesObj = sObj.optJSONObject("episodes") ?: return@forEach
                val parsedEpisodes = mutableListOf<AllohaEpisode>()

                episodesObj.keys().forEach { eKey ->
                    val eObj = episodesObj.getJSONObject(eKey)
                    val transObj = eObj.optJSONObject("translation") ?: return@forEach
                    val parsedTrans = mutableListOf<AllohaTranslation>()

                    transObj.keys().forEach { tKey ->
                        val tData = transObj.getJSONObject(tKey)
                        val transName = tData.optString("translation", "Unknown")
                        val iframe = tData.optString("iframe", "")
                        if (iframe.isNotBlank()) {
                            parsedTrans.add(
                                AllohaTranslation(
                                    id = tKey,
                                    name = transName,
                                    iframeUrl = iframe,
                                )
                            )
                        }
                    }
                    val seasonNum = sKey.toIntOrNull() ?: return@forEach
                    val episodeNum = eKey.toIntOrNull() ?: return@forEach
                    parsedEpisodes.add(
                        AllohaEpisode(
                            season = seasonNum,
                            episode = episodeNum,
                            translations = parsedTrans.sortedBy { it.name },
                        )
                    )
                }
                val seasonNum = sKey.toIntOrNull() ?: return@forEach
                parsedSeasons.add(
                    AllohaSeason(
                        season = seasonNum,
                        episodes = parsedEpisodes.sortedBy { it.episode },
                    )
                )
            }

            val sortedSeasons = parsedSeasons.sortedBy { it.season }
            Log.d(TAG, "fetchByKpId: returning ${sortedSeasons.size} seasons")
            AllohaApiResult(
                title = title,
                isSerial = true,
                movie = null,
                seasons = sortedSeasons,
            )
        } else {
            val iframe = dataObj.optString("iframe", "")
            Log.d(TAG, "fetchByKpId: movie iframe=$iframe")
            AllohaApiResult(
                title = title,
                isSerial = false,
                movie = if (iframe.isNotBlank()) {
                    AllohaMovie(
                        title = title,
                        iframeUrl = iframe,
                        translations = listOf(
                            AllohaTranslation(
                                id = "default",
                                name = title,
                                iframeUrl = iframe,
                            )
                        ),
                    )
                } else null,
                seasons = emptyList(),
            )
        }
    }
}

private suspend fun <T> retry(times: Int, delayMs: Long, block: suspend () -> T): T {
    var lastError: Throwable? = null
    repeat(times) { attempt ->
        try { return block() } catch (e: Throwable) {
            lastError = e
            if (attempt < times - 1) delay(delayMs)
        }
    }
    throw lastError!!
}

File: SlooshIOS/Core/Network/Alloha/AllohaRepository.swift
import Foundation

class AllohaRepository {
    private let token = "ffbd312217e27c4245f2678afe1881"
    private let session: URLSession

    init() {
        let configuration = URLSessionConfiguration.default
        // We might need to disable some cache or handle certs, but default is fine for api.alloha.tv usually
        self.session = URLSession(configuration: configuration)
    }

    func getContent(kpId: Int) async throws -> AllohaContent {
        let urlString = "https://api.alloha.tv/?token=\(token)&kp=\(kpId)"
        guard let url = URL(string: urlString) else {
            throw URLError(.badURL)
        }

        let (data, _) = try await session.data(from: url)
        
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dataObj = json["data"] as? [String: Any] else {
            throw URLError(.cannotParseResponse)
        }

        let title = dataObj["name"] as? String ?? "Unknown"
        
        if let seasonsObj = dataObj["seasons"] as? [String: Any] {
            var seasons: [AllohaSeason] = []
            
            for (sKey, sValue) in seasonsObj {
                guard let sObj = sValue as? [String: Any],
                      let episodesObj = sObj["episodes"] as? [String: Any] else { continue }
                
                var episodes: [AllohaEpisode] = []
                
                for (eKey, eValue) in episodesObj {
                    guard let eObj = eValue as? [String: Any],
                          let transObj = eObj["translation"] as? [String: Any] else { continue }
                    
                    var translations: [AllohaTranslation] = []
                    
                    for (tKey, tValue) in transObj {
                        guard let tData = tValue as? [String: Any],
                              let iframe = tData["iframe"] as? String,
                              !iframe.isEmpty else { continue }
                        
                        translations.append(AllohaTranslation(
                            id: tKey,
                            name: tData["translation"] as? String ?? "Unknown",
                            iframeUrl: iframe
                        ))
                    }
                    
                    if !translations.isEmpty {
                        episodes.append(AllohaEpisode(
                            num: Int(eKey) ?? 0,
                            translations: translations.sorted { $0.name < $1.name }
                        ))
                    }
                }
                
                if !episodes.isEmpty {
                    seasons.append(AllohaSeason(
                        num: Int(sKey) ?? 0,
                        episodes: episodes.sorted { $0.num < $1.num }
                    ))
                }
            }
            
            return AllohaContent(
                title: title,
                isSerial: true,
                movieIframe: nil,
                seasons: seasons.sorted { $0.num < $1.num }
            )
        } else {
            let iframe = dataObj["iframe"] as? String ?? ""
            return AllohaContent(
                title: title,
                isSerial: false,
                movieIframe: iframe.isEmpty ? nil : iframe,
                seasons: []
            )
        }
    }
}
